### ISSUE TITLE
Training loss becomes NaN after 250 iterations

### ISSUE THREAD

**Issue by L-Hugh** [LukeForeverYoung/UReader#4]
Hello,

I'm facing an issue with the training process where the loss becomes NaN after about 250 iterations. Here are my configurations:

gradient_accumulation_steps: 32
micro_batch_size: 1
Precision: bf16
Number of GPUs: 8

Here's a snippet of the relevant training logs:
```
{'loss': 4.2243, 'learning_rate': 1.2e-05, 'epoch': 0.0}
{'loss': 2.9407, 'learning_rate': 3.2000000000000005e-05, 'epoch': 0.0}
{'loss': 2.1148, 'learning_rate': 5.2000000000000004e-05, 'epoch': 0.0}
{'loss': 1.3225, 'learning_rate': 7.2e-05, 'epoch': 0.0}
{'loss': 1.5604, 'learning_rate': 9.200000000000001e-05, 'epoch': 0.0}
{'loss': 1.5617, 'learning_rate': 9.999908368967624e-05, 'epoch': 0.0}
{'loss': 1.1551, 'learning_rate': 9.99975565058033e-05, 'epoch': 0.0}
{'loss': 1.3678, 'learning_rate': 9.999602932193036e-05, 'epoch': 0.0}
{'loss': 1.4489, 'learning_rate': 9.999450213805742e-05, 'epoch': 0.0}
{'loss': 1.1685, 'learning_rate': 9.99929749541845e-05, 'epoch': 0.0}
{'loss': 1.4941, 'learning_rate': 9.999144777031155e-05, 'epoch': 0.0}
{'loss': 1.3755, 'learning_rate': 9.998992058643862e-05, 'epoch': 0.0}
{'loss': 1.1424, 'learning_rate': 9.998839340256567e-05, 'epoch': 0.0}
{'loss': 1.4246, 'learning_rate': 9.998686621869274e-05, 'epoch': 0.0}
{'loss': 1.5329, 'learning_rate': 9.998533903481979e-05, 'epoch': 0.0}
{'loss': 1.2079, 'learning_rate': 9.998381185094686e-05, 'epoch': 0.0}
{'loss': 1.146, 'learning_rate': 9.998228466707392e-05, 'epoch': 0.0}
{'loss': 1.253, 'learning_rate': 9.998075748320098e-05, 'epoch': 0.0}
{'loss': 1.1522, 'learning_rate': 9.997923029932804e-05, 'epoch': 0.0}
{'loss': 1.2506, 'learning_rate': 9.99777031154551e-05, 'epoch': 0.0}
{'loss': 1.235, 'learning_rate': 9.997617593158218e-05, 'epoch': 0.0}
{'loss': 1.2043, 'learning_rate': 9.997464874770923e-05, 'epoch': 0.0}
{'loss': 1.1899, 'learning_rate': 9.99731215638363e-05, 'epoch': 0.0}
{'loss': 1.3401, 'learning_rate': 9.997159437996335e-05, 'epoch': 0.0}
{'loss': 1.1811, 'learning_rate': 9.997006719609041e-05, 'epoch': 0.0}
{'loss': nan, 'learning_rate': 9.996854001221747e-05, 'epoch': 0.0}
{'loss': nan, 'learning_rate': 9.996701282834455e-05, 'epoch': 0.0}
{'loss': nan, 'learning_rate': 9.99654856444716e-05, 'epoch': 0.0}
{'loss': nan, 'learning_rate': 9.996395846059867e-05, 'epoch': 0.0}
{'loss': nan, 'learning_rate': 9.996258399511301e-05, 'epoch': 0.0}
{'loss': nan, 'learning_rate': 9.996105681124008e-05, 'epoch': 0.0}
{'loss': nan, 'learning_rate': 9.995952962736715e-05, 'epoch': 0.0}
{'loss': nan, 'learning_rate': 9.99580024434942e-05, 'epoch': 0.01}
```

Any insights on this would be greatly appreciated. Has anyone else encountered this or have suggestions for resolving it?

Thanks in advance.

**Comment by 1359347500cwc** [comment]
> Hello,
> 
> I'm facing an issue with the training process where the loss becomes NaN after about 250 iterations. Here are my configurations:
> 
> gradient_accumulation_steps: 32 micro_batch_size: 1 Precision: bf16 Number of GPUs: 8
> 
> Here's a snippet of the relevant training logs:
> 
> ```
> {'loss': 4.2243, 'learning_rate': 1.2e-05, 'epoch': 0.0}
> {'loss': 2.9407, 'learning_rate': 3.2000000000000005e-05, 'epoch': 0.0}
> {'loss': 2.1148, 'learning_rate': 5.2000000000000004e-05, 'epoch': 0.0}
> {'loss': 1.3225, 'learning_rate': 7.2e-05, 'epoch': 0.0}
> {'loss': 1.5604, 'learning_rate': 9.200000000000001e-05, 'epoch': 0.0}
> {'loss': 1.5617, 'learning_rate': 9.999908368967624e-05, 'epoch': 0.0}
> {'loss': 1.1551, 'learning_rate': 9.99975565058033e-05, 'epoch': 0.0}
> {'loss': 1.3678, 'learning_rate': 9.999602932193036e-05, 'epoch': 0.0}
> {'loss': 1.4489, 'learning_rate': 9.999450213805742e-05, 'epoch': 0.0}
> {'loss': 1.1685, 'learning_rate': 9.99929749541845e-05, 'epoch': 0.0}
> {'loss': 1.4941, 'learning_rate': 9.999144777031155e-05, 'epoch': 0.0}
> {'loss': 1.3755, 'learning_rate': 9.998992058643862e-05, 'epoch': 0.0}
> {'loss': 1.1424, 'learning_rate': 9.998839340256567e-05, 'epoch': 0.0}
> {'loss': 1.4246, 'learning_rate': 9.998686621869274e-05, 'epoch': 0.0}
> {'loss': 1.5329, 'learning_rate': 9.998533903481979e-05, 'epoch': 0.0}
> {'loss': 1.2079, 'learning_rate': 9.998381185094686e-05, 'epoch': 0.0}
> {'loss': 1.146, 'learning_rate': 9.998228466707392e-05, 'epoch': 0.0}
> {'loss': 1.253, 'learning_rate': 9.998075748320098e-05, 'epoch': 0.0}
> {'loss': 1.1522, 'learning_rate': 9.997923029932804e-05, 'epoch': 0.0}
> {'loss': 1.2506, 'learning_rate': 9.99777031154551e-05, 'epoch': 0.0}
> {'loss': 1.235, 'learning_rate': 9.997617593158218e-05, 'epoch': 0.0}
> {'loss': 1.2043, 'learning_rate': 9.997464874770923e-05, 'epoch': 0.0}
> {'loss': 1.1899, 'learning_rate': 9.99731215638363e-05, 'epoch': 0.0}
> {'loss': 1.3401, 'learning_rate': 9.997159437996335e-05, 'epoch': 0.0}
> {'loss': 1.1811, 'learning_rate': 9.997006719609041e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.996854001221747e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.996701282834455e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.99654856444716e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.996395846059867e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.996258399511301e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.996105681124008e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.995952962736715e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.99580024434942e-05, 'epoch': 0.01}
> ```
> 
> Any insights on this would be greatly appreciated. Has anyone else encountered this or have suggestions for resolving it?
> 
> Thanks in advance.

> Hello,
> 
> I'm facing an issue with the training process where the loss becomes NaN after about 250 iterations. Here are my configurations:
> 
> gradient_accumulation_steps: 32 micro_batch_size: 1 Precision: bf16 Number of GPUs: 8
> 
> Here's a snippet of the relevant training logs:
> 
> ```
> {'loss': 4.2243, 'learning_rate': 1.2e-05, 'epoch': 0.0}
> {'loss': 2.9407, 'learning_rate': 3.2000000000000005e-05, 'epoch': 0.0}
> {'loss': 2.1148, 'learning_rate': 5.2000000000000004e-05, 'epoch': 0.0}
> {'loss': 1.3225, 'learning_rate': 7.2e-05, 'epoch': 0.0}
> {'loss': 1.5604, 'learning_rate': 9.200000000000001e-05, 'epoch': 0.0}
> {'loss': 1.5617, 'learning_rate': 9.999908368967624e-05, 'epoch': 0.0}
> {'loss': 1.1551, 'learning_rate': 9.99975565058033e-05, 'epoch': 0.0}
> {'loss': 1.3678, 'learning_rate': 9.999602932193036e-05, 'epoch': 0.0}
> {'loss': 1.4489, 'learning_rate': 9.999450213805742e-05, 'epoch': 0.0}
> {'loss': 1.1685, 'learning_rate': 9.99929749541845e-05, 'epoch': 0.0}
> {'loss': 1.4941, 'learning_rate': 9.999144777031155e-05, 'epoch': 0.0}
> {'loss': 1.3755, 'learning_rate': 9.998992058643862e-05, 'epoch': 0.0}
> {'loss': 1.1424, 'learning_rate': 9.998839340256567e-05, 'epoch': 0.0}
> {'loss': 1.4246, 'learning_rate': 9.998686621869274e-05, 'epoch': 0.0}
> {'loss': 1.5329, 'learning_rate': 9.998533903481979e-05, 'epoch': 0.0}
> {'loss': 1.2079, 'learning_rate': 9.998381185094686e-05, 'epoch': 0.0}
> {'loss': 1.146, 'learning_rate': 9.998228466707392e-05, 'epoch': 0.0}
> {'loss': 1.253, 'learning_rate': 9.998075748320098e-05, 'epoch': 0.0}
> {'loss': 1.1522, 'learning_rate': 9.997923029932804e-05, 'epoch': 0.0}
> {'loss': 1.2506, 'learning_rate': 9.99777031154551e-05, 'epoch': 0.0}
> {'loss': 1.235, 'learning_rate': 9.997617593158218e-05, 'epoch': 0.0}
> {'loss': 1.2043, 'learning_rate': 9.997464874770923e-05, 'epoch': 0.0}
> {'loss': 1.1899, 'learning_rate': 9.99731215638363e-05, 'epoch': 0.0}
> {'loss': 1.3401, 'learning_rate': 9.997159437996335e-05, 'epoch': 0.0}
> {'loss': 1.1811, 'learning_rate': 9.997006719609041e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.996854001221747e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.996701282834455e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.99654856444716e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.996395846059867e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.996258399511301e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.996105681124008e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.995952962736715e-05, 'epoch': 0.0}
> {'loss': nan, 'learning_rate': 9.99580024434942e-05, 'epoch': 0.01}
> ```
> 
> Any insights on this would be greatly appreciated. Has anyone else encountered this or have suggestions for resolving it?
> 
> Thanks in advance.

I encountered the same problem. My loss was nan at the beginning of training.

**Comment by 1359347500cwc** [comment]
![截屏2023-11-07 16 53 38](https://github.com/LukeForeverYoung/UReader/assets/57097883/00c8ef81-1574-4aea-b6cd-03cfb1665555)
Have you solved this problem? Or do you have any clues? I tried mPlug-owl's solution, but it didn't work at all.
![Uploading 截屏2023-11-07 16.56.36.png…]()

**Comment by LukeForeverYoung** [comment]
> ![截屏2023-11-07 16 53 38](https://user-images.githubusercontent.com/57097883/280976798-00c8ef81-1574-4aea-b6cd-03cfb1665555.png) Have you solved this problem? Or do you have any clues? I tried mPlug-owl's solution, but it didn't work at all. ![截屏2023-11-07 16 56 36](https://user-images.githubusercontent.com/57097883/280978469-85e90416-003c-4db7-a7f1-052a44fa7449.png)

I have uncommented the loss_mask in the forward function. This resolved the NaN issue that was occurring in my experiments.
